### ISSUE TITLE
[quantize_fp8_with_scale] compiled kernel silently reused when input shape changes. Shape guard does not invalidate stale kernel

### ISSUE THREAD

**Issue by DevAyush22** [torch-spyre/torch-spyre#2523]
### What happened?

When torch.compile() is applied to **quantize_fp8_with_scale** and the compiled
function is called first with shape (2,8), then with (4,8), (5,8) etc. in the
same process, the kernel compiled for (2,8) is silently reused for all subsequent
shapes. The (2,8) kernel only writes 2 rows. All extra rows in the new shape
remain as zeros. No error or warning is raised.
 
The shape guard in the Spyre inductor backend does not correctly detect that the
input shape has changed and does not recompile.

### What did you expect to happen?

When the input shape changes, the shape guard must detect the change and
recompile the kernel for the new shape. Every row of the output must contain
correct FP8-quantized values matching the CPU reference.

### How can we reproduce it?

    import torch
    import torch_spyre
 
    fn    = torch.compile(
        lambda x, s: torch.ops.spyre.quantize_fp8_with_scale(x, s),
        backend="inductor"
    )
    scale = torch.tensor(1.0, dtype=torch.float16, device="spyre")
 
    # First call: compiles and caches kernel for shape (2,8)
    out1 = fn(torch.full((2, 8), 2.0, dtype=torch.float16, device="spyre"), scale)
    print(out1.cpu().float().tolist())
    # [[2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
    #  [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0]]  correct
 
    # Second call: different shape (3,8) -- reuses stale (2,8) kernel
    out2 = fn(torch.full((3, 8), 2.0, dtype=torch.float16, device="spyre"), scale)
    print(out2.cpu().float().tolist())
    # [[2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
    #  [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
    #  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]  BUG: row 2 should be 2.0

### Any environmental details we need to know?

Platform: Linux, Python 3.12.12, PyTorch 2.12.0
Device:   Spyre
dtype:    torch.float16 input, torch.float8_e4m3fn output
Mode:     torch.compile(backend="inductor")

### Anything else we need to know?

Related to Parent Issue #2276
Found while testing PR #2401 

Confirmed across multiple shape transitions (both increasing and decreasing M):
 
  Seed (2,8), then call without reset:
    (2,8)  PASS
    (4,8)  FAIL: rows 2-3 zeros  (16/32 wrong)
    (5,8)  FAIL: rows 2-4 zeros  (24/40 wrong)
    (8,8)  FAIL: rows 2-7 zeros  (48/64 wrong)
    M=3,5,7 (odd M): also FAIL -- not a hardware constraint, purely cache reuse
 
  Seed (8,8), then call without reset:
    (8,8)  FAIL, (4,8) FAIL, (2,8) PASS  -- works in reverse direction too
 
  torch._dynamo.reset() is NOT a reliable workaround:
    Hardware binary cache (dxp_standalone) persists beyond Python-level reset.
    The same shape (2,24) produced PASS in one test and FAIL in another
    within the same script run, both with reset() called before each call.
    There is no reliable user-space workaround.

Any production inference code that reuses a compiled quantize_fp8_with_scale
function across different batch sizes silently produces wrong FP8 values
with no error. This corrupts all downstream computations silently.

### Relevant log output

```Shell
# First call (2,8) -- correct:
    [[2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
     [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0]]
 
    # Second call (3,8) without reset -- stale kernel reused:
    [[2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
     [2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0, 2.0],
     [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]  <- row 2 all zeros, expected 2.0
```

**Comment by DevAyush22** [comment]
cc: @Akash-Nayak @joyalbin @pradghos

**Comment by DevAyush22** [comment]
The original repro script called .cpu() directly on the raw FP8 quantize output, which now fails with a separate unrelated error (see #3023).

Updated repro uses the full quantize→dequantize roundtrip so the output is FP16:

```
import torch
import torch_spyre  # noqa: F401

FP16    = torch.float16
FP32    = torch.float32
FP8_MAX = 448.0
DEVICE  = "spyre"

def roundtrip(x, scale):
    q = torch.ops.spyre.quantize_fp8_with_scale(x, scale)
    return torch.ops.spyre.dequantize_fp8_with_scale(q, scale)

def make_input(shape):
    n = 1
    for d in shape: n *= d
    return ((torch.arange(n, dtype=FP32) % 16) + 1).to(FP16).reshape(shape).to(DEVICE)

def cpu_ref(x_cpu, scale_cpu):
    inv = 1.0 / scale_cpu.to(FP16)
    fp8 = (x_cpu.to(FP16) * inv).clamp(-FP8_MAX, FP8_MAX).to(torch.float8_e4m3fn)
    return fp8.to(FP16) * scale_cpu.to(FP16)

fn    = torch.compile(roundtrip, backend="inductor")
scale = torch.tensor(1.0, dtype=FP16, device=DEVICE)

x2   = make_input((2, 8))
out2 = fn(x2, scale)
print(f"(2, 8): {'PASS' if torch.allclose(out2.cpu(), cpu_ref(x2.cpu(), scale.cpu()), atol=0.0, rtol=0.0) else 'FAIL'}")

x3   = make_input((3, 8))
out3 = fn(x3, scale)
print(f"(3, 8): {'PASS' if torch.allclose(out3.cpu(), cpu_ref(x3.cpu(), scale.cpu()), atol=0.0, rtol=0.0) else 'FAIL'}")
```

**Output:**

```
(2, 8): PASS
(3, 8): PASS
```

Issue is not observed with the latest stack. Shape guard recompile now correctly generates a new kernel for each new shape. Thus, closing this issue.
